client/src/main/java/org/apache/celeborn/client/write/DataPushQueue.java

```diff
@@ -0,0 +1,102 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.celeborn.client.write;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.celeborn.client.ShuffleClient;
+import org.apache.celeborn.common.CelebornConf;
+import org.apache.celeborn.common.protocol.PartitionLocation;
+import org.apache.celeborn.common.util.Utils;
+
+public class DataPushQueue {
+  private static final Logger logger = LoggerFactory.getLogger(DataPushQueue.class);
+
+  private final List<LinkedBlockingQueue<PushTask>> workingQueuePerPartition;
+  private final PushState pushState;
+  private final DataPusher dataPusher;
+
+  private final String appId;
+  private final int shuffleId;
+  private final int numMappers;
+  private final int numPartitions;
+  private final ShuffleClient client;
+  private int partitionIdIdx = 0;
+
+  public DataPushQueue(
+      CelebornConf conf,
+      DataPusher dataPusher,
+      ShuffleClient client,
+      String appId,
+      int shuffleId,
+      int mapId,
+      int attemptId,
+      int numMappers,
+      int numPartitions) {
+    this.appId = appId;
+    this.shuffleId = shuffleId;
+    this.numMappers = numMappers;
+    this.numPartitions = numPartitions;
+    this.client = client;
+    this.dataPusher = dataPusher;
+    final String mapKey = Utils.makeMapKey(shuffleId, mapId, attemptId);
+    this.pushState = client.getOrRegisterPushState(mapKey);
+    workingQueuePerPartition = new ArrayList<>(numPartitions);
+    final int capacity = conf.pushQueueCapacity();
+    for (int i = 0; i < numPartitions; i++) {
+      workingQueuePerPartition.add(new LinkedBlockingQueue<>(capacity));
+    }
+  }
+
+  public LinkedBlockingQueue<PushTask> takeAnyWorkingQueue() throws IOException {
+    while (!dataPusher.terminatedOrHasException()) {
+      int partitionId = nextPartitionId();
+      LinkedBlockingQueue<PushTask> pushTasks = workingQueuePerPartition.get(partitionId);
+      if (!pushTasks.isEmpty()) {
+        Map<Integer, PartitionLocation> partitionLocationMap =
+            client.getOrRegisterShuffle(appId, shuffleId, numMappers, numPartitions);
+        PartitionLocation loc = partitionLocationMap.get(partitionId);
+        boolean reachLimit = pushState.limitMaxInFlight(loc.hostAndPushPort());
+        if (!reachLimit) {
+          return pushTasks;
+        }
+      }
+    }
+    return null;
+  }
+
+  public LinkedBlockingQueue<PushTask> takeWorkingQueue(int partitionId) {
+    return workingQueuePerPartition.get(partitionId);
+  }
+
+  public void clear() {
+    workingQueuePerPartition.parallelStream().forEach(LinkedBlockingQueue::clear);
+    workingQueuePerPartition.clear();
+  }
+
+  private int nextPartitionId() {
+    return partitionIdIdx++ % numPartitions;
```

Comment:
Don't understand why use `nextPartitionId` like this, can you explain more?

Reply:
nextPartitionId, it will loop get PushTask from workingQueuePerPartition. 
If one worker is reach limitation, it will skip this partition and get it next loop.